constantine/math/arithmetic/finite_fields_precomp_square_root.nim

```diff
@@ -0,0 +1,198 @@
+  acc.prod(acc, z255)
+  square_repeated(acc, 8)
+  acc.prod(acc, z255)
+  square_repeated(acc, 6)
+  acc.prod(acc, z11)
+  square_repeated(acc, 9)
+  acc.prod(acc, z255)
+  square_repeated(acc, 2)
+  acc.prod(acc, z)
+  square_repeated(acc, 7)
+  acc.prod(acc, z255)
+  square_repeated(acc, 8)
+  acc.prod(acc, z255)
+  square_repeated(acc, 8)
+  acc.prod(acc, z255)
+  square_repeated(acc, 8)
+  acc.prod(acc, z255)
+  # acc is now z^((BaseFieldMultiplicativeOddOrder - 1)/2)
+  rootOfUnity.square(acc)
+  rootOfUnity.prod(rootOfUnity ,z)
+  squareRootCandidate.prod(acc, z)
+
+
+func invSqrtEqDyadic*(z: var Fp): bool =
+  ## The algorithm works by essentially computing the dlog of z and then halving it.
+  ## negExponent is intended to hold the negative of the dlog of z.
+  ## We determine this 32-bit value (usually) _sqrtBlockSize many bits at a time, starting with the least-significant bits.
+  ##
+  ## If _sqrtBlockSize does not divide 32, the *first* iteration will determine fewer bits.
+
+  var negExponent: int
+  var temp, temp2: Fp
+
+  # set powers[i] to z^(1<< (i*blocksize))
+  var powers: array[4, Fp]
+  powers[0] = z
+  for i in 1..<Fp.C.tonelliShanks(sqrtParam_Blocks):
+    powers[i] = powers[i - 1]
+    for j in 0..<Fp.C.tonelliShanks(sqrtParam_BlockSize):
+      powers[i].square(powers[i])
+
+  ## looking at the dlogs, powers[i] is essentially the wanted exponent, left-shifted by i*_sqrtBlockSize and taken mod 1<<32
+  ## dlogHighDyadicRootNeg essentially (up to sign) reads off the _sqrtBlockSize many most significant bits. (returned as low-order bits)
+  ## 
+  ## first iteration may be slightly special if BlockSize does not divide 32
+  negExponent = sqrtAlg_NegDlogInSmallDyadicSubgroup(powers[Fp.C.tonelliShanks(sqrtParam_Blocks) - 1])
+  negExponent = negExponent shr Fp.C.tonelliShanks(sqrtParam_FirstBlockUnusedBits)
+
+  # if the exponent we just got is odd, there is no square root, no point in determining the other bits
+  if (negExponent and 1) == 1:
+    return false
+
+  for i in 1..<Fp.C.tonelliShanks(sqrtParam_Blocks):
+    temp2 = powers[Fp.C.tonelliShanks(sqrtParam_Blocks) - 1 - i]
+    for j in 0..<i:
+      sqrtAlg_GetPrecomputedRootOfUnity(temp, int( (negExponent shr (j*Fp.C.tonelliShanks(sqrtParam_BlockSize))) and Fp.C.tonelliShanks(sqrtParam_BitMask) ), uint(j + Fp.C.tonelliShanks(sqrtParam_Blocks) - 1 - i))
+      temp2.prod(temp2, temp)
+    
+    var newBits = sqrtAlg_NegDlogInSmallDyadicSubgroup(temp2)
+    negExponent = negExponent or (newBits shl ((i*Fp.C.tonelliShanks(sqrtParam_BlockSize)) - Fp.C.tonelliShanks(sqrtParam_FirstBlockUnusedBits)))
+
+  negExponent = negExponent shr 1
+  z.setOne()
+
+  for i in 0..<Fp.C.tonelliShanks(sqrtParam_Blocks):
+    sqrtAlg_GetPrecomputedRootOfUnity(temp, int((negExponent shr (i*Fp.C.tonelliShanks(sqrtParam_BlockSize))) and Fp.C.tonelliShanks(sqrtParam_BitMask)), uint(i))
+    z.prod(z, temp)
+
+  return true
+
+func sqrtPrecomp*(dst: var Fp, x: Fp): SecretBool {.inline.} =
```

Comment:
This function is not necessary, `invSqrt` is the more universal primitive and what should be exported